



The ipcoal ethos 
================

Welcome to *ipcoal*, a simple tool for simulating and analyzing genealogies and sequences across genomes in a phylogenetic context. *ipcoal* builds upon the powerful coalescent simulator `msprime <https://msprime.readthedocs.io/en/stable/>`__ to automate the task of describing complex demographic models (e.g., divergence times, effective population sizes, migration events) when simulating the coalescent process on phylogenetic trees or networks. 

.. *ipcoal* can parse and visualize trees using the emphasizes the difference between true genealogical variation across genomes in comparison to gene trees that can be inferred from sequence alignments. 

The *ipcoal* package is designed for use in jupyter-notebooks to encourage reproducible and shareable results. It is a powerful tool for posterior predictive data analysis, methods validation, for generating training data for machine learning, and as a teaching tool for learning the coalescent process in an interactive and visual environment.


Our goals
---------
- Accessible: Easy to install and easy to use, with documentation and cookbooks.
- Reproducible: Designed for interactive use in Jupyter notebooks.
- Reliable: Tested on each build.
- Comprehendable: We provide simple methods to verify and visualize models.
- Didactic: We provide (binder) notebooks for teaching in a classroom.
- Transparent: Read our code and make it your own.



Contact us
-------------
- Questions? Email us: p.mckenzie@columbia.edu
- Feature request? Raise a ticket on `github <http://github.com/eaton-lab/ipcoal>`__.